src/vsr/batch.zig

```diff
@@ -0,0 +1,804 @@
+                assert(BatchDecoder.init(
+                    self.context,
+                    buffer[0..bytes_written],
+                ) != error.BatchInvalid);
+            }
+
+            return bytes_written;
+        }
+    };
+}
+
+const test_batch = struct {
+    const Context = struct {
+        const empty: Context = .{};
+    };
+    const Operation = enum(u8) {
+        size_2 = constants.vsr_operations_reserved + 1,
+        size_4 = constants.vsr_operations_reserved + 2,
+        size_8 = constants.vsr_operations_reserved + 3,
+        size_16 = constants.vsr_operations_reserved + 4,
+        size_32 = constants.vsr_operations_reserved + 5,
+        size_64 = constants.vsr_operations_reserved + 6,
+        size_128 = constants.vsr_operations_reserved + 7,
+        size_256 = constants.vsr_operations_reserved + 8,
+
+        fn to_vsr(operation: Operation) vsr.Operation {
+            return @enumFromInt(@intFromEnum(operation));
+        }
+
+        fn from_vsr(operation: vsr.Operation) Operation {
+            return @enumFromInt(@intFromEnum(operation));
+        }
+    };
+
+    fn ElementType(comptime operation: Operation) type {
+        const size = Adapter.element_size(
+            Context.empty,
+            operation.to_vsr(),
+        );
+
+        const alignment = Adapter.alignment(
+            Context.empty,
+            operation.to_vsr(),
+        );
+
+        const Element = extern struct {
+            value: [size]u8 align(alignment),
+        };
+        assert(@sizeOf(Element) == size);
+        assert(@alignOf(Element) == alignment);
+
+        return Element;
+    }
+
+    const Adapter = struct {
+        fn is_valid(context: Context, vsr_operation: vsr.Operation) bool {
+            _ = context;
+            if (vsr_operation.vsr_reserved()) return false;
+            const pow = @intFromEnum(vsr_operation) - vsr.constants.vsr_operations_reserved;
+            return pow >= 1 and pow <= 8; // Test sizes between 2^1 and 2^8.
+        }
+
+        fn element_size(context: Context, vsr_operation: vsr.Operation) usize {
+            assert(is_valid(context, vsr_operation));
+            const pow = @intFromEnum(vsr_operation) - vsr.constants.vsr_operations_reserved;
+            return std.math.pow(usize, 2, pow);
+        }
+
+        fn alignment(context: Context, vsr_operation: vsr.Operation) usize {
+            assert(is_valid(context, vsr_operation));
+            const operation: Operation = @enumFromInt(@intFromEnum(vsr_operation));
+            return switch (operation) {
+                .size_2 => @alignOf(u16),
+                .size_4 => @alignOf(u32),
+                .size_8 => @alignOf(u64),
+                .size_16 => @alignOf(u128),
+                .size_32 => @alignOf(u32),
+                .size_64 => @alignOf(u64),
+                .size_128 => @alignOf(u128),
+                .size_256 => @alignOf(u128),
+            };
+        }
+    };
+
+    fn run(options: struct {
```

Comment:
Rewrote this function here since I had too many suggestions and wanted to experiment: https://github.com/tigerbeetle/tigerbeetle/commit/6d6492c2e7c6ab3f044e61b0fb33feb2ca62489d

Primary goal is to not rely on `BatchEncoder` when generating the data to be encoded. We are testing encoder, so we shouldn't trust it!